src/rate_limiters.rs

```diff
@@ -0,0 +1,234 @@
+/// let rate_limiter = PerEventCap::new(100);
+/// let rate_limiter_fn = Arc::new(move |msg| rate_limiter.should_allow(msg));
+/// 
+/// // Register the rate limiter
+/// // app.set_rate_limiter(rate_limiter_fn);
+/// ```
+pub struct PerEventCap {
+    events_per_minute: u32,
+    event_counters: DashMap<String, EventCounter>,
+}
+
+#[derive(Debug)]
+struct EventCounter {
+    count: u32,
+    window_start: Instant,
+}
+
+impl EventCounter {
+    fn new() -> Self {
+        Self {
+            count: 0,
+            window_start: Instant::now(),
+        }
+    }
+
+    fn reset_if_expired(&mut self) {
+        let now = Instant::now();
+        if now.duration_since(self.window_start) >= Duration::from_secs(60) {
+            self.count = 0;
+            self.window_start = now;
+        }
+    }
+
+    fn increment(&mut self) {
+        self.count += 1;
+    }
+
+    fn current_count(&self) -> u32 {
+        self.count
+    }
+}
+
+impl PerEventCap {
+    /// Create a new PerEventCap rate limiter
+    /// 
+    /// # Arguments
+    /// * `events_per_minute` - Maximum number of events allowed per minute for each event type
+    pub fn new(events_per_minute: u32) -> Self {
+        Self {
+            events_per_minute,
+            event_counters: DashMap::new(),
+        }
+    }
+
+    /// Check if an event should be allowed based on the rate limit
+    /// Returns true if the event should be sent, false if it should be dropped
+    pub fn should_allow(&self, message: &rudderanalytics::message::Message) -> bool {
+        let event_type = self.extract_event_type(message);
+        
+        // Use entry API to get or insert a new counter
+        let mut counter = self.event_counters.entry(event_type).or_insert_with(EventCounter::new);
+        
+        // Reset counter if the time window has expired
+        counter.reset_if_expired();
+        
+        // Check if we're within the limit
+        if counter.current_count() < self.events_per_minute {
+            counter.increment();
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Extract the event type from a RudderStack message
+    fn extract_event_type(&self, message: &rudderanalytics::message::Message) -> String {
+        match message {
+            rudderanalytics::message::Message::Track(track) => {
+                track.event.clone()
+            }
+            rudderanalytics::message::Message::Identify(_) => "identify".to_string(),
+            rudderanalytics::message::Message::Page(page) => {
+                page.name.clone()
+            }
+            rudderanalytics::message::Message::Screen(screen) => {
+                screen.name.clone()
+            }
+            rudderanalytics::message::Message::Group(_) => "group".to_string(),
+            rudderanalytics::message::Message::Alias(_) => "alias".to_string(),
+            rudderanalytics::message::Message::Batch(_) => "batch".to_string(),
+        }
+    }
```

Comment:
i believe this should work but havnt run it
```suggestion
    fn extract_event_type<'a>(&self, message: &'a rudderanalytics::message::Message) -> &'a str {
        match message {
            rudderanalytics::message::Message::Track(track) => {
                &track.event
            }
            rudderanalytics::message::Message::Identify(_) => "identify",
            rudderanalytics::message::Message::Page(page) => {
                &page.name
            }
            rudderanalytics::message::Message::Screen(screen) => {
                &screen.name
            }
            rudderanalytics::message::Message::Group(_) => "group",
            rudderanalytics::message::Message::Alias(_) => "alias",
            rudderanalytics::message::Message::Batch(_) => "batch",
        }
    }
```

Reply:
this one won't work

```
error[E0507]: cannot move out of `screen.name` which is behind a shared reference
   --> src/rate_limiters.rs:101:17
    |
101 |                 screen.name
    |                 ^^^^^^^^^^^ move occurs because `screen.name` has type `std::string::String`, which does not implement the `Copy` trait
```

etc

Reply:
update this now works fine